components/Modal/ManageSources.js

```diff
@@ -16,104 +29,487 @@ const messages = defineMessages({
+  },
+  check_gpg: {
+    defaultMessage: "Check GPG key"
+  },
+  selectOne: {
+    defaultMessage: "Select one"
+  },
+  typeRepo: {
+    defaultMessage: "yum repository"
+  },
+  typeMirrorlist: {
+    defaultMessage: "mirrorlist"
+  },
+  typeMetalink: {
+    defaultMessage: "metalink"
+  },
+  add: {
+    defaultMessage: "Add Source"
+  },
+  save: {
+    defaultMessage: "Add Source"
+  },
+  update: {
+    defaultMessage: "Update Source"
+  },
+  cancel: {
+    defaultMessage: "Cancel"
   }
 });
 
 class ManageSources extends React.Component {
-  constructor() {
-    super();
+  constructor(props) {
+    super(props);
+    this.state = {
+      showModal: false
+    };
+    this.open = this.open.bind(this);
+    this.close = this.close.bind(this);
+  }
+
+  open() {
+    this.setState({ showModal: true });
+  }
+
+  close() {
+    this.setState({ showModal: false });
+  }
+
+  render() {
+    return (
+      <>
+        <a href="#" onClick={this.open}>
+          <FormattedMessage
+            defaultMessage="Manage Sources"
+            description="User action for displaying the list of source repositories"
+          />
+        </a>
+        {this.state.showModal && (
+          <ManageSourcesModal
+            manageSources={this.props.manageSources}
+            addSource={this.props.addModalManageSourcesEntry}
+            removeSource={this.props.removeModalManageSourcesEntry}
+            clearError={this.props.modalManageSourcesFailure}
+            close={this.close}
+            intl={this.props.intl}
+          />
+        )}
+      </>
+    );
+  }
+}
+
+class ManageSourcesModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      addEntry: false,
+      url: "",
+      name: "",
+      type: "",
+      check_ssl: false,
+      check_gpg: false,
+      warningDuplicateName: false,
+      warningDuplicateUrl: false,
+      editName: ""
+    };
+    this.handleShowAddEntry = this.handleShowAddEntry.bind(this);
+    this.handleChange = this.handleChange.bind(this);
+    this.handleValidateName = this.handleValidateName.bind(this);
+    this.handleValidateUrl = this.handleValidateUrl.bind(this);
+    this.handleEditSource = this.handleEditSource.bind(this);
+    this.handleSubmitSource = this.handleSubmitSource.bind(this);
+  }
+
+  componentDidUpdate(prevProps) {
+    // if no errors are returned on add/edit, then reset form state after fetching sources completes
+    if (
+      Object.keys(this.props.manageSources.error).length === 0 &&
+      !this.props.manageSources.fetchingSources &&
+      prevProps.manageSources.fetchingSources
+    ) {
+      this.setState({
+        addEntry: false,
+        url: "",
+        name: "",
+        type: "",
+        check_ssl: false,
+        check_gpg: false
+      });
+    }
+  }
+
+  handleShowAddEntry(e, state) {
+    this.setState({ addEntry: state });
+    if (state) {
+      this.setState({
+        editName: ""
```

Comment:
I pass this to the SourcesListItem so that if this value matches the props in SourcesListItem, then focus is placed on the Edit button. I.e. the user clicks Edit to show the form, and when the form closes, focus is returned to the Edit button they clicked.

Let me know if there's a better way to handle focus, or better labels to make it more obvious that's what's happening.